From 4446abd2635b58f2d25a0f10b1d50da12c1aba70 Mon Sep 17 00:00:00 2001
From: Robert Mader <robert.mader@collabora.com>
Date: Wed, 28 Jan 2026 13:49:26 +0100
Subject: [PATCH 4/6] aperture: Set video/image caps to I420 and NV12

The used format can influence the resulting video profile. Notably RGB
formats from the libcamera SW/GPU-ISP would, when used with x264enc, be
converted to Y444, i.e. a non-subsampled format, resulting in a H264
High (4/4/4) profile being used - which is undesirable for various
reasons.

Ensure we use I420, which is the default format for SDR content, and
only allow NV12 as alternative for HW-encoding.

(cherry picked from commit 444e8ebb94fde9d2b17b7d67ea0166d4c1b0cc42)
---
 aperture/src/viewfinder.rs | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/aperture/src/viewfinder.rs b/aperture/src/viewfinder.rs
index bfeb0e0..9fd6f02 100644
--- a/aperture/src/viewfinder.rs
+++ b/aperture/src/viewfinder.rs
@@ -315,11 +315,25 @@ mod imp {
                 .expect("Missing GStreamer Base Plug-ins");
             camerabin.set_property("video-filter", &videoconvert_video);
 
+            let caps_video = gst_video::video_make_raw_caps(&[
+                gst_video::VideoFormat::I420,
+                gst_video::VideoFormat::Nv12,
+            ])
+            .build();
+            camerabin.set_property("video-capture-caps", caps_video);
+
             let videoconvert_image = gst::ElementFactory::make("videoconvert")
                 .build()
                 .expect("Missing GStreamer Base Plug-ins");
             camerabin.set_property("image-filter", &videoconvert_image);
 
+            let caps_image = gst_video::video_make_raw_caps(&[
+                gst_video::VideoFormat::I420,
+                gst_video::VideoFormat::Nv12,
+            ])
+            .build();
+            camerabin.set_property("image-capture-caps", caps_image);
+
             self.sink_paintable.set(paintablesink).unwrap();
 
             self.picture
-- 
2.52.0

